src/tools/rbd_mirror/image_map/Policy.cc

```diff
@@ -0,0 +1,684 @@
+
+      need_map_update = true;
+      m_image_action.init_state(global_image_id);
+      m_image_info.emplace(global_image_id, Info());
+    }
+
+    m_image_info[global_image_id].registered_peers.insert(mirror_uuid);
+    if (need_map_update) {
+      schedule_add_action(global_image_id, instance_id);
+    }
+  }
+}
+
+template <typename I>
+void Policy<I>::schedule_remove_action(const std::string &global_image_id,
+                                       const std::string &instance_id,
+                                       std::set<std::string> &&unregistered_peers) {
+  dout(20) << ": global_image_id=" << global_image_id << ", instance_id="
+           << instance_id << dendl;
+
+  ImageAction::Action action(ImageAction::ACTION_TYPE_REMOVE);
+  action.on_unmap = new FunctionContext(
+    [this, global_image_id, instance_id](int r) {
+      assert(m_map_lock.is_wlocked());
+      dout(20) << ": action_instance_id=" << instance_id
+               << " for global_image_id=" << global_image_id << dendl;
+
+      m_image_info[global_image_id].action_instance_id = instance_id;
+    });
+
+  action.on_purge = new FunctionContext(
+    [this, global_image_id, unregistered_peers](int r) {
+      assert(m_map_lock.is_wlocked());
+      dout(20) << ": action_peers count=" << unregistered_peers.size()
+               << " for global_image_id=" << global_image_id << dendl;
+
+      m_image_info[global_image_id].action_peers = unregistered_peers;
+    });
+
+  action.on_finish = new FunctionContext(
+    [this, global_image_id, instance_id](int r) {
+      assert(m_map_lock.is_wlocked());
+      if (r < 0) {
+        derr << ": failed to remove global_image_id=" << global_image_id
+             << " from instance_id=" << instance_id << dendl;
+      }
+
+      m_image_info[global_image_id].action_instance_id.clear();
+      m_image_info[global_image_id].action_peers.clear();
+
+      // if no actions are pending, do away with the state
+      if (!m_image_action.actions_pending(global_image_id)) {
+        m_image_info.erase(global_image_id);
+        m_image_action.release_state(global_image_id);
+      };
+    });
+
+  append_and_schedule_action(global_image_id, action);
+}
+
+template <typename I>
+void Policy<I>::update_images_removed(const std::string &mirror_uuid,
+                                      const std::set<std::string> &global_image_ids) {
+  dout(20) << dendl;
+  assert(m_map_lock.is_wlocked());
+
+  for (auto const &global_image_id : global_image_ids) {
+    std::string instance_id = lookup(global_image_id);
+    assert(instance_id != UNMAPPED_INSTANCE_ID);
+
+    auto rm = m_image_info[global_image_id].registered_peers.erase(mirror_uuid);
+    assert(rm);
+
+    if (!mirror_uuid.empty()) {
+      m_image_info[global_image_id].unregistered_peers.insert(mirror_uuid);
+    }
+
+    if (m_image_info[global_image_id].registered_peers.empty()) {
+      unmap(global_image_id);
+      schedule_remove_action(global_image_id, instance_id,
```

Comment:
Just checking back on this -- we do not need to wait for (all) removal notification acks from the mirror daemon peer before scheduling an `unmap` action, right? We'll be firing off the removal notifications, and on the last remove (empty peers) schedule a remove action (`unmap + purge`). This would maintain the order or `remove` and `release`.

Reply:
Assuming ```registered_peers``` contains both the remote peers and the local "peer", there shouldn't be a need to wait. If the last peer is removed it implies that the local image has successfully been deleted.

Reply:
yup, thats what it is.